soccer/planning/SettlePathPlanner.cpp

```diff
@@ -0,0 +1,507 @@
+    ShapeSet obstaclesWBall = obstacles;
+    obstaclesWBall.add(make_shared<Circle>(ball.pos, *_ballAvoidDistance));
+
+    // Previous RRT path from last iteration
+    std::unique_ptr<Path>& prevPath = planRequest.prevPath;
+
+    // The small beginning part of the previous path
+    unique_ptr<Path> partialPath = nullptr;
+
+    // The path is from the original robot position to the intercept point
+    // We only care about the replan lead time from the current pos in the path
+    // to the intercept point
+    RJ::Seconds timeIntoPreviousPath;
+
+    // How much of the future we are devoting to the partial path
+    // 0ms unless we have a partial path, then it's partialReplanLeadTime
+    RJ::Seconds partialPathTime = 0ms;
+
+
+    // Max angle from the ball vector when trying to bounce the ball
+    // on dampen when trying to speed up actions after captures
+    const float maxAngleoffBallForDampen = 45; // deg
+
+    // Check and see if we should reset the entire thing if we are super far off course
+    // or the ball state changes significantly
+    checkSolutionValidity(ball);
+
+
+    // Smooth out the ball velocity a little bit so we can get a better estimate of intersect points
+    if (firstBallVelFound) {
+        averageBallVel = applyLowPassFilter<Point>(averageBallVel, ball.vel, *_ballVelGain);
+    } else {
+        averageBallVel = ball.vel;
+        firstBallVelFound = true;
+    }
+
+    // Change start instant to be the partial path end instead of the robot current location
+    // if we actually have already calculated a path the frame before
+    if (prevPath) {
+        // How much of the previous path to steal
+        const RJ::Seconds partialReplanLeadTime = RRTPlanner::getPartialReplanLeadTime();
+
+        timeIntoPreviousPath = curTime - prevPath->startTime();
+
+        // Make sure we still have time in the path to replan and correct
+        // since it's likely that the old path is slightly off
+        //
+        // ---|----------------|-----------------|
+        // TimeNow     EndPartialPrevPath  FinalTargetPoint
+        //                     |-----------------|
+        //          Amount of the path we can change this iteration
+        if (timeIntoPreviousPath < prevPath->getDuration() - 2*partialReplanLeadTime &&
+            timeIntoPreviousPath > 0ms) {
+            partialPath =
+                prevPath->subPath(0ms, timeIntoPreviousPath + partialReplanLeadTime);
+            partialPathTime = partialPath->getDuration() - timeIntoPreviousPath;
+            startInstant = partialPath->end().motion;
+        }
+    }
+
+    // Check if we should transition from intercept to dampen
+    // Start instant may be changed in that case since we want to start changing the path
+    // as soon as possible
+    processStateTransition(ball, prevPath.get(), timeIntoPreviousPath, startInstant);
+
+    // Run state code
+    // TODO: Do I need to do a move or is it implicit?
```

Comment:
This actually can't do Return value optimization since there's multiple return paths. But function calls return temporary values anyways. THe std:move just casts values to temporaries so it's basically a no-op to have it here.

Reply:
By temporaries I mean rvalue.